### PR TITLE
Break up --hub-router into discovery and relay

### PR DESCRIPTION
**Pull request by nerdalert**
- Details in issue #610
- Example output in the PR comments.

### CODE REVIEW

**Comment by nerdalert** [comment]
The following is output from 5 nodes that covers the peering/hole-punching scenarios we support with this patch. **Note:** Not sure if adding a field warrants a SQL column to Device or if editing an existing migration is how we want to handle field changes:


### 100.100.0.1 Discovery Node - EC2 open 51820

- Peers to all node EXCEPT hard NAT/symmetric NAT nodes

```
sudo ./nexd-dev-linux-amd64 --username kitteh1 --password floofykittens --stun -discovery-node https://try.nexodus.local

ubuntu@ip-172-31-83-131:~$ sudo wg show wg0 dump
aIs9sKKOs0OBxM1Oy1ANPvu0t+BHkTTW4sCIOHXzVEA=	9D9Rr26/rbscwKzNJ7iFTG1U3WW16r4itR2AC0NUvkE=	51820	off
itc587APYo5VxKmjBRZjh8LCJSFDgM95PfDUNrcxsmo=	(none)	44.211.198.0:51820	100.100.0.0/16	1679183363	91932	93780	off
fGjIwqk8rzvuxqEZVghrvDbKU1wKpR2pNFkD2++UXG0=	(none)	54.167.97.135:40101	100.100.0.3/32	1679183286	102304	101076	off
ps4lyG5gX76IuqRoRQn7L8kpI55ajrFimsIMivE9+lM=	(none)	71.29.252.138:62603	100.100.0.4/32	1679183365	97696	96832	off
```

### 100.100.0.2 Relay Node - EC2 open 51820

- Peers to all nodes. This is from the same VPC as the disco node with 51820 open. I think this port opening is required for hard NAT nodes but need to validate.

```
sudo ./nexd-dev-linux-amd64 --username kitteh1 --password floofykittens --stun -relay-node https://try.nexodus.local


ubuntu@ip-172-31-91-6:~$ sudo wg show wg0 dump
IJPsILZCzDx4CZM3VI5NoEXBkCt8648SZER3PuAe8HA=	itc587APYo5VxKmjBRZjh8LCJSFDgM95PfDUNrcxsmo=	51820	off
9D9Rr26/rbscwKzNJ7iFTG1U3WW16r4itR2AC0NUvkE=	(none)	44.203.54.250:51820	100.100.0.1/32	1679183363	99668	97820	off
fGjIwqk8rzvuxqEZVghrvDbKU1wKpR2pNFkD2++UXG0=	(none)	54.167.97.135:40101	100.100.0.3/32	1679183426	97492	93596	off
ps4lyG5gX76IuqRoRQn7L8kpI55ajrFimsIMivE9+lM=	(none)	71.29.252.138:62603	100.100.0.4/32	1679183393	150492	148244	off
aoUKqyXCEiuj+GXKMPPasHpo1msB8m3ADa9FKbsOil4=	(none)	71.29.252.138:61232	100.100.0.5/32	1679183413	106700	107140	off
```

### 100.100.0.3 EC2 Node  - no inbound ports open to the VPC

- Peers to all nodes except the symmetric NAT node which it reaches via the relay

```
sudo ./nexd-dev-linux-arm64 --username kitteh1 --password floofykittens https://try.nexodus.local


ubuntu@ip-10-180-15-110:~$ sudo wg show wg0 dump
MCcbq0hiM12c2W+4/sroNSSCMt1D9hohK1Vtfm1cmFs=	fGjIwqk8rzvuxqEZVghrvDbKU1wKpR2pNFkD2++UXG0=	40101	off
9D9Rr26/rbscwKzNJ7iFTG1U3WW16r4itR2AC0NUvkE=	(none)	44.203.54.250:51820	100.100.0.1/32	1679183655	116564	118204	off
itc587APYo5VxKmjBRZjh8LCJSFDgM95PfDUNrcxsmo=	(none)	44.211.198.0:51820	100.100.0.0/16	1679183673	110420	114620	off
ps4lyG5gX76IuqRoRQn7L8kpI55ajrFimsIMivE9+lM=	(none)	71.29.252.138:62603	100.100.0.4/32	1679183632	111864	113712	off
```


### 100.100.0.4 OSX Node easy nat - remote edge site, no inbound ports open

- Peers with all nodes - it peers to the symmetric NAT node (192.168.1.254:2499) since it is a local node in the same remote site.

```
sudo ./nexd --username kitteh1 --password floofykittens https://try.nexodus.local

sudo wg show utun8 dump
ANp3SoGswzxABRcbpkpTZwI57AOxk32vjTXxyHCDl0k=	ps4lyG5gX76IuqRoRQn7L8kpI55ajrFimsIMivE9+lM=	62603	off
9D9Rr26/rbscwKzNJ7iFTG1U3WW16r4itR2AC0NUvkE=	(none)	44.203.54.250:51820	100.100.0.1/32	1679184102	134248	136408	off
itc587APYo5VxKmjBRZjh8LCJSFDgM95PfDUNrcxsmo=	(none)	44.211.198.0:51820	100.100.0.0/16	1679184003	198680	201512	off
fGjIwqk8rzvuxqEZVghrvDbKU1wKpR2pNFkD2++UXG0=	(none)	54.167.97.135:40101	100.100.0.3/32	1679183996	135972	134984	off
aoUKqyXCEiuj+GXKMPPasHpo1msB8m3ADa9FKbsOil4=	(none)	192.168.1.254:2499	100.100.0.5/32	1679184033	76496	75620	off
```

### 100.100.0.5 Symmetric NAT (behind Pfsense hard NAT)  - remote edge site, no inbound ports open

- 2 Peers, one is a local node sharing the same reflexive address and the other is the relay, it does not peer to the discovery node since it is symmetric.

```
sudo ./nexd-dev-linux-arm64 --username kitteh1 --password floofykittens https://try.nexodus.local

ubuntu@aml-s905x-cc:~$ sudo wg show wg0 dump
0Lrh1AkOirlkU0JH7sv3HbONV8G6arcTP4vyYTwM73A=	aoUKqyXCEiuj+GXKMPPasHpo1msB8m3ADa9FKbsOil4=	34212	off
itc587APYo5VxKmjBRZjh8LCJSFDgM95PfDUNrcxsmo=	(none)	44.211.198.0:51820	100.100.0.0/16	1679183413	110084	111052	off
ps4lyG5gX76IuqRoRQn7L8kpI55ajrFimsIMivE9+lM=	(none)	192.168.1.9:62603	100.100.0.4/32	1679183422	43268	43852	off
```

File: internal/nexodus/nexodus.go [hunk 1]
```diff
@@ -252,13 +255,30 @@ func (ax *Nexodus) Start(ctx context.Context, wg *sync.WaitGroup) error {
...
+			}
+		}
+
+		if ax.discoveryNode {
+			existingDiscoveryNode, err := ax.orgDiscoveryCheck(peerListing)
+			if err != nil {
+				return err
+			}
+			if existingDiscoveryNode != uuid.Nil {
+				return fmt.Errorf("the organization already contains a discovery node, device %s needs to be deleted before adding a new relay", existingDiscoveryNode)
```
Comment (vishnoianil) [inline internal/nexodus/nexodus.go:280]: "...before adding a new **discovery node**"

File: internal/nexodus/nexodus.go [hunk 2]
```diff
@@ -593,3 +607,25 @@ func (ax *Nexodus) orgRelayCheck() (uuid.UUID, error) {
 
 	return relayID, nil
 }
+
+// orgRelayCheck checks if there is an existing relay in the organization that does not match this devices pub key
```
Comment (vishnoianil) [inline internal/nexodus/nexodus.go:611]: comment needs fixiing.

**Review by vishnoianil — APPROVED** [review]
minor log/comments fixing is needed, otherwise looks good to me.

**Comment by vishnoianil** [comment]
@nerdalert apex-connectivity.md is now noxodus-connectivity.md, so it needs a rebase.

**Pull request merged**